appcues/src/main/java/com/appcues/ui/ModalStateMachineOwner.kt

```diff
@@ -0,0 +1,167 @@
+package com.appcues.ui
+
+import android.view.MotionEvent
+import android.view.View
+import android.view.ViewTreeObserver
+import android.view.ViewTreeObserver.OnDrawListener
+import android.view.ViewTreeObserver.OnGlobalFocusChangeListener
+import android.view.ViewTreeObserver.OnGlobalLayoutListener
+import android.view.ViewTreeObserver.OnScrollChangedListener
+import android.view.ViewTreeObserver.OnTouchModeChangeListener
+import android.view.ViewTreeObserver.OnWindowAttachListener
+import android.view.ViewTreeObserver.OnWindowFocusChangeListener
+import com.appcues.AppcuesCoroutineScope
+import com.appcues.data.model.RenderContext
+import com.appcues.data.model.RenderContext.Modal
+import com.appcues.monitor.AppcuesActivityMonitor
+import com.appcues.statemachine.Action.Retry
+import com.appcues.statemachine.Error.StepError
+import com.appcues.statemachine.StateMachine
+import com.appcues.ui.utils.getParentView
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import java.util.Timer
+import java.util.TimerTask
+import kotlin.concurrent.schedule
+
+internal class ModalStateMachineOwner(
+    override val stateMachine: StateMachine,
+    private val coroutineScope: AppcuesCoroutineScope,
+) : StateMachineOwning {
+
+    companion object {
+        private const val SCROLL_DEBOUNCE_MILLISECONDS = 1_000L
+    }
+
+    override val renderContext: RenderContext = Modal
+
+    // used to attempt retry when items in view change
+    private var viewTreeUpdateHandler: ViewTreeUpdateHandler = ViewTreeUpdateHandler { attemptRetry() }
+    private var uiIdleDebounceTimer: TimerTask? = null
+    private var viewTreeObserver: ViewTreeObserver? = null
+
+    init {
+        // set up observer on the Modal state machine errors to initiate retry
+        coroutineScope.launch(Dispatchers.IO) {
+            stateMachine.errorFlow.collect { error ->
+                // if a recoverable step error is observed, begin attempting retry
+                if (error is StepError && error.recoverable) {
+                    startRetryHandling()
+                }
+            }
+        }
+    }
+
+    override suspend fun reset() {
+        stateMachine.stop(true)
+    }
+
+    fun stopRetryHandling() {
+        viewTreeUpdateHandler.detach()
+        viewTreeObserver = null
+    }
+
+    private fun startRetryHandling() {
+        if (viewTreeObserver == null) {
+            AppcuesActivityMonitor.activity?.getParentView()?.viewTreeObserver?.let {
+                viewTreeUpdateHandler.attach(it)
+                this.viewTreeObserver = it
+            }
+        }
+    }
+
+    private fun attemptRetry() {
+        onUiIdle {
+            // cancel any current touch (drag) as our overlay is about to drop on top and we don't want things moving behind it
+            AppcuesActivityMonitor.activity?.getParentView()?.dispatchTouchEvent(
+                MotionEvent.obtain(0L, 0L, MotionEvent.ACTION_CANCEL, 0f, 0f, 0)
+            )
```

Comment:
this needs to happen _before_ the retry, since it might cause pager views to "snap" into place, and the tooltip needs to be positioned based on the final position of the view, not the position prior to that auto-snap animation.